data/tutorials/lg_08_error_handling.md

```diff
@@ -125,67 +191,549 @@ let () =
+
+Using `Option` it is possible to write functions that return `None` instead of
+throwing an exception.
+```ocaml
+let div_opt m n =
+  try Some (m / n) with
+    Division_by_zero -> None
+
+let find_opt p l =
+  try Some (List.find p l) with
+    Not_found -> None
+```
+We can try those functions:
+
+```ocaml
+# 1 / 0;;
+Exception: Division_by_zero.
+# div_opt 42 2;;
+- : int option = Some 24
+# div_opt 42 0;;
+- : int option = None
+# List.find (fun x -> x mod 2 = 0) [1; 3; 5];;
+Exception: Not_found.
+# find_opt (fun x -> x mod 2 = 0) [1; 3; 4; 5];;
+- : int option = Some 4
+# find_opt (fun x -> x mod 2 = 0) [1; 3; 5];;
+- : int option = None
+```
+
+This can even be turned into a higher-order generic function:
+```ocaml
+# let try_opt f x = try Some (f x) with _ -> None;;
+val try_opt : ('a -> 'b) -> 'a -> 'b option = <fun>
+```
 
-## Result Type
+It tends to be considered good practice nowadays when a function can fail in
+cases that are not bugs (i.e., not `assert false`, but network failures, keys
+not present, etc.) to return a more explicit type such as `'a option` or `('a,
+'b) result` (see next section).
 
-The Stdlib module contains the following type:
+### Naming Conventions
+
+There are two naming conventions to have two versions of the same partial
+function: one raising exception, the other returning an option. In the above
+examples, the convention of the standard library is to add an `_opt` suffix to
+name of the version of the function that returns an option instead of raising
+exceptions.
+```ocaml
+val find: ('a -> bool) -> 'a list -> 'a
+(** @raise Not_found *)
+val find_opt: ('a -> bool) -> 'a list -> 'a option
+```
+This is extracted from the `List` module of the standard library.
+
+However, some projects tend to avoid or reduce the usage of exceptions. In such
+a context, reversing the convention is a relatively common idiom. It is the
+version of the function that raises exceptions that is suffixed with `_exn`.
+Using the same functions, that would be the specification
+```ocaml
+val find_exn: ('a -> bool) -> 'a list -> 'a
+(** @raise Not_found *)
+val find: ('a -> bool) -> 'a list -> 'a option
+```
+### Composing Functions Returning Options
+
+The function `div_opt` can't raise exceptions. However, since it doesn't return
+a result of type `int`, it can't be used in place of an `int`. The same way
+OCaml doesn't
+[promote](https://en.wikipedia.org/wiki/Type_conversion#Type_promotion) integers
+into floats, it doesn't automatically converts `int option` into `int` or _vice
+versa_.
+
+```ocaml
+# 21 + Some 21;;
+Error: This expression has type 'a option
+       but an expression was expected of type int
+```
+
+In order to combine option values with other values, conversion functions are
+needed. Here are the functions provided by the `Option` module to extract the
+data contained in an option:
+```ocaml
+val get : 'a t -> 'a
+val value : 'a t -> default:'a -> 'a
+val fold : none:'a -> some:('b -> 'a) -> 'b t -> 'a
```

Comment:
I suspect all of these examples are meant for information not an example to run when following this document, and I suspect that will be clear to the end reader. :)